plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/gitlabCreateOrMerge.ts

```diff
@@ -0,0 +1,317 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { resolveSafeChildPath } from '@backstage/backend-common';
+import { Config } from '@backstage/config';
+import { InputError } from '@backstage/errors';
+import { ScmIntegrationRegistry } from '@backstage/integration';
+import { Types } from '@gitbeaker/core';
+import { Gitlab } from '@gitbeaker/node';
+import path from 'path';
+import { serializeDirectoryContents } from '../../../../lib/files';
+import { createTemplateAction } from '../../createTemplateAction';
+import { initRepoAndPush } from '../helpers';
+import { getRepoSourceDirectory, parseRepoUrl } from './util';
+
+/**
+ * Creates a new action that initializes a git repository of the content in the workspace
+ * and publishes it to GitLab.
+ *
+ * @public
+ */
+export function createGitlabCreateOrMergeAction(options: {
```

Comment:
Just a discussion point: Should we keep two actions for GitLab (or in general any SCM)?

IMO this action is better given it checks with GitLab whether the report provided exists or not and then it does either of the forward-moving action:

1. Raise a GitLab repo if repoURL doesn't point to existing repository
2. Raise an MR against the existing repository if repoUrl is valid.

Outside the scope of this Pull Request:

Personally, we should generally not have a string input for Owner field as its manual entry and error-prone (coming from experience of user feedback) - it would be much better if we can make an API request to Git SCM and return an array of namespaces the signed-in user is allowed to create repository in and let him/her select from that menu.